beacon-chain/core/light-client/store.go

```diff
@@ -41,6 +43,120 @@ func NewLightClientStore(db iface.HeadAccessDatabase, p p2p.Accessor, e event.Su
+	blockRoot, err := attestedBlock.Block().HashTreeRoot()
+	if err != nil {
+		return errors.Wrapf(err, "failed to compute attested block root")
+	}
+	parentRoot := [32]byte(update.AttestedHeader().Beacon().ParentRoot)
+	signatureBlockRoot, err := block.Block().HashTreeRoot()
+	if err != nil {
+		return errors.Wrapf(err, "failed to compute signature block root")
+	}
+	headRoot, err := s.beaconDB.HeadBlockRoot()
+	if err != nil {
+		return errors.Wrapf(err, "failed to get head block root from database")
+	}
+	newBlockIsHead := signatureBlockRoot == headRoot
+
+	// create the new cache item
+	cacheItem := &lightClientCacheItem{
+		period: period,
+		slot:   block.Block().Slot(),
+	}
+
+	// check if parent exists in cache
+	parentItem, ok := s.nonFinalityCache.items[parentRoot]
+	if !ok {
+		// if not, create an item for the parent, but don't need to save it.
+		bestUpdateSoFar, err := s.beaconDB.LightClientUpdate(ctx, period)
+		if err != nil {
+			return errors.Wrapf(err, "could not get best light client update for period %d", period)
+		}
+		parentItem = &lightClientCacheItem{
+			period:             period,
+			bestUpdate:         &bestUpdateSoFar,
+			bestFinalityUpdate: &s.lastFinalityUpdate,
+		}
+	} else {
+		cacheItem.parent = parentItem
+	}
+
+	// if at a period boundary, no need to compare data, just save new ones
+	if parentItem.period != period {
+		cacheItem.bestUpdate = &update
+		cacheItem.bestFinalityUpdate = &finalityUpdate
+		s.nonFinalityCache.items[blockRoot] = cacheItem
+
+		s.setLastOptimisticUpdate(optimisticUpdate, true)
+
+		// if the new block is not head, we don't want to change our lastFinalityUpdate
+		if newBlockIsHead {
+			s.setLastFinalityUpdate(finalityUpdate, true)
+		}
+
+		return nil
+	}
+
+	// if in the same period, compare updates
+	isUpdateBetter, err := IsBetterUpdate(update, *parentItem.bestUpdate)
+	if err != nil {
+		return errors.Wrapf(err, "could not compare light client updates")
+	}
+	if isUpdateBetter {
+		cacheItem.bestUpdate = &update
+	} else {
+		cacheItem.bestUpdate = parentItem.bestUpdate
+	}
+
+	var finalityUpdateChanged bool
+	isBetterFinalityUpdate := IsBetterFinalityUpdate(finalityUpdate, *parentItem.bestFinalityUpdate)
+	if isBetterFinalityUpdate {
+		finalityUpdateChanged = true
+		cacheItem.bestFinalityUpdate = &finalityUpdate
+	} else {
+		finalityUpdateChanged = false
+		cacheItem.bestFinalityUpdate = parentItem.bestFinalityUpdate
+	}
+
+	// save new item in cache
+	s.nonFinalityCache.items[blockRoot] = cacheItem
+
+	// save lastOptimisticUpdate if better
+	if isBetterOptimisticUpdate := IsBetterOptimisticUpdate(optimisticUpdate, s.lastOptimisticUpdate); isBetterOptimisticUpdate {
+		s.setLastOptimisticUpdate(optimisticUpdate, true)
+	}
+
+	// if the new block is considered the head, set the last finality update
+	if newBlockIsHead {
+		s.setLastFinalityUpdate(*cacheItem.bestFinalityUpdate, finalityUpdateChanged)
```

Comment:
You don't really need the `finalityUpdateChanged` variable, you can pass `isBetterFinalityUpdate`